pkg/distributor/distributor.go

```diff
@@ -448,6 +454,28 @@ func (d *Distributor) Push(ctx context.Context, req *logproto.PushRequest) (*log
 	var validationErrors util.GroupedErrors
 	validationContext := d.validator.getValidationContextForTime(time.Now(), tenantID)
 
+	shardStreamsCfg := d.validator.Limits.ShardStreams(tenantID)
+	maybeShardByRate := func(stream logproto.Stream, pushSize int) {
+		if shardStreamsCfg.Enabled {
+			streams = append(streams, d.shardStream(stream, pushSize, tenantID)...)
+		} else {
+			streams = append(streams, KeyedStream{
+				HashKey: lokiring.TokenFor(tenantID, stream.Labels),
+				Stream:  stream,
+			})
+		}
+	}
+	maybeShardByTime := func(stream logproto.Stream, labels labels.Labels, pushSize int) {
+		if shardStreamsCfg.TimeShardingEnabled {
+			streamsByTime := shardStreamByTime(stream, labels, d.ingesterCfg.MaxChunkAge/2)
```

Comment:
I would apply sharding by time only if the logs are older than `now - d.ingesterCfg.MaxChunkAge/2` 
It would let the `normal` logs to be ingested without the need to create a time-bucket every hour. At the same time, it would still allow old logs to be ingested without strict order requirements. 
wdyt?

Reply:
I believe it would help us to avoid such spikes of created streams 
<img width="1261" alt="image" src="https://github.com/user-attachments/assets/8b6e9d58-e331-45ad-a6fc-a3af6f6458d5">

Reply:
I considered this, but I'm somewhat worried that it might cause more issues than it solves:
1. Assuming that the number of out-of-order logs is not insignificant, this will actually result in more total and more active streams than the current approach. We will have logs in both `{foo="bar"}` and `{foo="bar", __time_shard="111_222"}` streams.
2. The "now" value of the distributors can be slightly different than the "now" value of the ingesters. And even if they are the same, some time passes until the ingesters handle the push request. So if we try to calculate in the distributors which logs will be rejected by the ingesters based on the current time, we might be wrong in a small percentage of the cases, resulting in TooFarBehind errors. We can add a safety margin, of course, but it would make everything more complicated. 

So, because this is a per-tenant config, I think the tradeoffs are slightly better if we always inject the `__time_shard__` label when it's enabled for a specific tenant, even for current data :thinking: Or maybe it should be configurable and we should support both ways? :thinking:  What do you think, am I missing something here?

Reply:
>  this will actually result in more total and more active streams than the current approach.

I do not think so. 
let say we ingest 4 logs lines with ts: `now` , `now-15m`, `now - 1h`, `now-1d`. 
With your version, we would split these logs to 4 separate streams with `__time_shard__` : `A`, `B`, `C`, `D`. 
(If `now` and `now-15m` are from different time buckets).

With the changes that I propose, we would not create new buckets for `now` , `now-15m`.  So, in result, we would have 3 streams: `original`, `C`, `D`.  



In general, I would try to guess if the data will be rejected by the ingester or not (maybe with some safe margin, maybe 15m is enough), because we would not create new streams, every hour, for the data that is fresh enough...

We already see that it almost doubles the stream count for the cases when a lot of fresh data is ingested. Also, it affects the chunks, because we get more underutilized chunks that are flushed due to the reason `idle`.

Also, this change does not give us any drawbacks in terms of ingesting out-of-order old logs...

Reply:
> With the changes that I propose, we would not create new buckets for `now` , `now-15m`. So, in result, we would have 3 streams: `original`, `C`, `D`.

The question is how are the streams going to be distributed over time. Because in your example, if you keep getting more and more logs that have `now`, `now-15m`, `now - 1h`, `now-1d` timestamps as the value of `now` changes, eventually you'd still cover the whole time range with 2 sets of streams - ones that do have `__time_shard__` and the original ones that don't. Which probably will be worse than the current approach.

But if these out-of-order logs arrive only occasionally, we'd be much better off with what you suggest, for sure.

So I think I'll add another per-tenant config option called something like `time_sharding_ignore_recent` with a value of 30 minutes or something like that. Any logs with timestamps greater than  `now  -time_sharding_ignore_recent`  won't be split into a `__time_shard__` stream, but if we set the option to `0`, everything will be sharded (i.e. the current behavior from this PR will be applied). It should give us the flexibility to configure this to achieve the optimal results for different tenants. WDYT?

Reply:
yep, it would be ideal.

Reply:
This should now be resolved by the most recent commit, PTAL: https://github.com/grafana/loki/pull/14711/commits/778773528be3b804c50567414df57673ac17daf1